google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1568,53 +1600,80 @@ public void run() {
         });
   }
 
+  /**
+   * Returns the minimum of the wanted number of sessions that the caller wants to create and the
+   * actual max number that may be created at this moment.
+   */
+  private int getAllowedCreateSessions(int wantedSessions) {
+    synchronized (lock) {
+      return Math.min(
+          wantedSessions, options.getMaxSessions() - (totalSessions() + numSessionsBeingCreated));
+    }
+  }
+
   private boolean canCreateSession() {
     synchronized (lock) {
       return totalSessions() + numSessionsBeingCreated < options.getMaxSessions();
     }
   }
 
-  private void createSession() {
-    logger.log(Level.FINE, "Creating session");
+  private void createSessions(final int sessionCount) {
+    logger.log(Level.FINE, String.format("Creating %d sessions", sessionCount));
     synchronized (lock) {
-      numSessionsBeingCreated++;
-      executor.submit(
-          new Runnable() {
-            @Override
-            public void run() {
-              SessionImpl session = null;
-              try {
-                session = spanner.createSession(db);
-                logger.log(Level.FINE, "Session created");
-              } catch (Throwable t) {
-                // Expose this to customer via a metric.
-                synchronized (lock) {
-                  numSessionsBeingCreated--;
-                  if (isClosed()) {
-                    decrementPendingClosures();
-                  }
-                  handleCreateSessionFailure(newSpannerException(t));
-                }
-                return;
-              }
-              boolean closeSession = false;
-              PooledSession pooledSession = null;
-              synchronized (lock) {
-                pooledSession = new PooledSession(session);
-                numSessionsBeingCreated--;
-                if (closureFuture != null) {
-                  closeSession = true;
-                } else {
-                  Preconditions.checkState(totalSessions() <= options.getMaxSessions() - 1);
-                  allSessions.add(pooledSession);
-                  releaseSession(pooledSession);
-                }
-              }
-              if (closeSession) {
-                closeSession(pooledSession);
-              }
-            }
-          });
+      numSessionsBeingCreated += sessionCount;
+      try {
+        // Create a batch of sessions. The actual session creation can be split into
+        // multiple gRPC calls and the sessions are returned by the enumerator as they come
```

Comment:
"the sessions are returned by the enumerator" should be "the session consumer consumes the returned sessions as they become available".